pyuvdata/uvdata.py

```diff
@@ -3514,158 +3672,305 @@ def read(self, filename, axis=None, file_type=None,
+                    antenna_names=antenna_names, ant_str=ant_str, bls=bls,
+                    frequencies=frequencies, freq_chans=freq_chans,
+                    times=times,
+                    polarizations=polarizations, blt_inds=blt_inds,
+                    read_data=read_data, run_check=run_check,
+                    check_extra=check_extra,
+                    run_check_acceptability=run_check_acceptability,
+                    data_array_dtype=data_array_dtype,
+                    keep_all_metadata=keep_all_metadata)
+
+                if select:
+                    unique_times = np.unique(self.time_array)
+                    select_times = unique_times[
+                        np.where((unique_times >= np.min(time_range))
+                                 & (unique_times <= np.max(time_range)))]
 
             if select:
-                unique_times = np.unique(self.time_array)
-                times_to_keep = unique_times[np.where((unique_times >= np.min(time_range))
-                                                      & (unique_times <= np.max(time_range)))]
-                self.select(times=times_to_keep, run_check=run_check, check_extra=check_extra,
-                            run_check_acceptability=run_check_acceptability,
-                            keep_all_metadata=keep_all_metadata)
+                if file_type in ['fhd', 'ms', 'mwa_corr_fits']:
+                    # these file types do not have select on read, so set all
+                    # select parameters
+                    select_antenna_nums = antenna_nums
+                    select_antenna_names = antenna_names
+                    select_ant_str = ant_str
+                    select_bls = bls
+                    select_frequencies = frequencies
+                    select_freq_chans = freq_chans
+                    select_times = times
+                    select_polarizations = polarizations
+                    select_blt_inds = blt_inds
+
+                    if time_range is not None:
+                        unique_times = np.unique(self.time_array)
+                        select_times = unique_times[
+                            np.where((unique_times >= np.min(time_range))
+                                     & (unique_times <= np.max(time_range)))]
+
+                self.select(
+                    antenna_nums=select_antenna_nums,
+                    antenna_names=select_antenna_names,
+                    ant_str=select_ant_str,
+                    bls=select_bls,
+                    frequencies=select_frequencies,
+                    freq_chans=select_freq_chans,
+                    times=select_times,
+                    polarizations=select_polarizations,
+                    blt_inds=select_blt_inds,
+                    run_check=run_check,
+                    check_extra=check_extra,
+                    run_check_acceptability=run_check_acceptability,
+                    keep_all_metadata=keep_all_metadata)
```

Comment:
Just popping in after looking at the tests. With the exception of `select_times` it looks like all of these select keywords only get defined if the file type is one of `fhd`, `ms`, or `mwa_corr_fits`. I am betting this is because other files types support select on read but we _also_ want to support the selection with the `time_range` keyword. Is it crazy to have two selects? One inside the if file type check for all the `select_meta_data` parameters, and one outside for the select times? Sounds like it might slow things down a little for large datasets though.

Reply:
In the other file types, if `time_range` is set then `select_times` is set and this select is run. There used to be separate selects for each file type but I consolidated them here.

Reply:
see e.g. line 3781

Reply:
just fixed this.